lib/hash_ring/ets.ex

```diff
@@ -287,83 +362,196 @@ defmodule ExHashRing.HashRing.ETS do
         {nil, pending_gcs} ->
           pending_gcs
 
         {_stale_timer_ref, pending_gcs} ->
-          do_ring_gen_gc(table, ring_gen)
+          :ok = do_ring_gen_gc(state.previous_table, ring_gen)
+          :ok = do_ring_gen_gc(state.current_table, ring_gen)
           pending_gcs
       end
 
-    {:noreply, %{state | pending_gcs: pending_gcs}}
+    {:noreply, %__MODULE__{state | pending_gcs: pending_gcs}}
   end
 
-  defp rebuild(
-         %{nodes: nodes, name: name, table: table, ring_gen: ring_gen, overrides: overrides} =
-           state
-       ) do
-    new_ring_gen = ring_gen + 1
+  ## Private
 
-    :ets.insert(
-      table,
-      for {hash, node} <- Utils.gen_items(nodes) do
-        {{new_ring_gen, hash}, node}
-      end
-    )
+  defp do_find_node(table, gen, num_nodes, key) do
+    hash = Utils.hash(key)
 
-    config =
-      if map_size(overrides) > 0 do
-        overrides =
-          overrides
-          |> Enum.filter(fn {_, values} -> length(values) > 0 end)
-          |> Map.new()
+    case find_next_highest_item(table, gen, num_nodes, hash) do
+      {_, node} ->
+        {:ok, node}
 
-        {table, new_ring_gen, length(nodes), overrides}
-      else
-        {table, new_ring_gen, length(nodes)}
+      _ ->
+        {:error, :invalid_ring}
+    end
+  end
+
+  @spec do_find_nodes_in_table(
+    key :: term(),
+    table :: :ets.tid(),
+    overrides :: override_map(),
+    gen :: generation(),
+    num_nodes :: non_neg_integer(),
+    num :: non_neg_integer()
+  ) :: {:ok, [binary]} | {:error, term}
+  defp do_find_nodes_in_table(_key, _table, _overrides, _gen, 0, _num) do
+    {:error, :invalid_ring}
+  end
+
+  defp do_find_nodes_in_table(_key, _table, _overrides, _gen, _num_nodes, 0) do
+    {:ok, []}
+  end
+
+  defp do_find_nodes_in_table(key, table, overrides, gen, num_nodes, num) when map_size(overrides) == 0 do
+    {:ok, do_find_nodes(table, gen, num_nodes, num, Utils.hash(key), [], 0)}
+  end
+
+  defp do_find_nodes_in_table(key, table, overrides, gen, num_nodes, num) do
+    {found, found_length} =
+      case overrides do
+        %{^key => overrides} ->
+          Utils.take_max(overrides, num)
+
+        _ ->
+          {[], 0}
       end
 
-    Config.set(name, self(), config)
+    {:ok, do_find_nodes(table, gen, num_nodes, max(num - found_length, 0), Utils.hash(key), found, found_length)}
+  end
 
-    schedule_ring_gen_gc(%{state | ring_gen: new_ring_gen}, ring_gen)
+  defp do_find_nodes(_table, _gen, _num_nodes, 0, _hash, found, _found_length) do
+    Enum.reverse(found)
   end
 
-  def schedule_ring_gen_gc(state, 0), do: state
+  defp do_find_nodes(_table, _gen, num_nodes, _remaining, _hash, found, num_nodes) do
+    Enum.reverse(found)
+  end
+
+  defp do_find_nodes(table, gen, num_nodes, remaining, hash, found, found_length) do
+    {number, node} = find_next_highest_item(table, gen, num_nodes, hash)
+
+    if node in found do
+      do_find_nodes(
```

Comment:
TBH with you this `do_find_nodes` recursion is rough for me to follow.
maybe the parameter `found` is too general? Or maybe the fact that there's no spec on it which makes me think it's a boolean.